Tweak binop errors

estebank opened the pull request:
* Suggest potentially missing binop trait bound (fix #73416)
* Use structured suggestion for dereference in binop

estebank committed:
Tweak binop errors

* Suggest potentially missing binop trait bound (fix #73416)
* Use structured suggestion for dereference in binop

rust-highfive commented:
r? @davidtwco

(rust_highfive has picked a reviewer for you, use r? to override)

davidtwco commented on src/librustc_typeck/check/op.rs:
```diff
@@ -318,18 +330,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                        sugg = true;
+                                    }
+                                }
+                                if !sugg && !suggested_deref {
```
Is `sugg` necessary, couldn't this have remained `else if !suggested_deref`?

estebank replied:
I was thinking that if the single type found by the visitor wasn't a type param we wouldn't show the suggestion, but I'm seeing that the visitor is only looking at type params so the if let would be irrefutable. Changing.

davidtwco commented on src/librustc_typeck/check/op.rs:
```diff
@@ -254,6 +256,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 if !lhs_ty.references_error() && !rhs_ty.references_error() {
```
Pre-existing, but couldn't we make this into a match arm to avoid indenting the next 300 lines?

```rust
Err(()) if lhs_ty.references_error() || rhs_ty.references_error() => self.tcx.ty_error(),
Err(()) => {
 // as before..
}
```

estebank replied:
Done

davidtwco commented on src/librustc_typeck/check/op.rs:
```diff
@@ -478,18 +514,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                        sugg = true;
+                                    }
+                                }
+                                if !sugg && !suggested_deref && !involves_fn {
```
As above, couldn't this remain an `else if`?

davidtwco commented on src/librustc_typeck/check/op.rs:
```diff
@@ -458,18 +485,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                                            rty.peel_refs(),
-                                            lstring
-                                        ));
+                                        err.span_suggestion_verbose(
```
A lot of this error reporting seems duplicated between the `IsAssign::No` and `IsAssign::Yes` arms - is there something I'm missing or could we move some of this into a function and just call it from both arms?

davidtwco commented on src/librustc_typeck/check/op.rs:
```diff
@@ -478,18 +514,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                    if let ty::Param(p) = ty.kind {
+                                        // FIXME: This *guesses* that constraining the type param
+                                        // will make the operation available, but this is only true
+                                        // when the corresponding trait has a blanked
```
```suggestion
                                        // when the corresponding trait has a blanket
```

davidtwco commented on src/librustc_typeck/check/op.rs:
```diff
@@ -318,18 +330,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                    if let ty::Param(p) = ty.kind {
+                                        // FIXME: This *guesses* that constraining the type param
+                                        // will make the operation available, but this is only true
+                                        // when the corresponding trait has a blanked
```
```suggestion
                                        // when the corresponding trait has a blanket
```

davidtwco commented on src/librustc_typeck/check/op.rs:
```diff
@@ -318,18 +330,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                        // implementation, like the following:
+                                        // `impl<'a> PartialEq for &'a [T] where T: PartialEq {}`
+                                        // The correct thing to do would be to verify this
+                                        // projection would hold.
```
Is there a reason that we cannot construct a predicate here and check if it holds?

estebank replied:
Laziness? 😅

estebank replied:
Done.

davidtwco reviewed (changes requested):
Error improvements look good, handful of comments.

estebank committed:
review comments: clean up code

* deduplicate logic
* fix typos
* remove unnecessary state

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=32849&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32849/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/32849/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux mingw-check
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 23'
Agent machine name: 'fv-az578'
Current agent version: '2.170.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200614.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200614.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.171.1)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/c8151a9c-fe8a-49df-bd41-c3b365f1f885.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73674/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73674/merge:refs/remotes/pull/73674/merge
---
 ---> a9ec21d337b3
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> 5ff2c13d8dba
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            python3 ../x.py doc --stage 0 src/libstd &&            /scripts/validate-toolstate.sh
 ---> 6b931e755c7e
Successfully built 6b931e755c7e
Successfully tagged rust-ci:latest
Built container sha256:6b931e755c7ea1f69816640eb9df74fafd40a545d0e5aa8341d35009dabb0f3c
---
    Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
    Checking chalk-engine v0.11.0
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
    Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
configure: dist.missing-tools   := True
configure: build.cargo-native-static := True
configure: llvm.ccache          := sccache
configure: rust.dist-src        := False
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Hugepagesize:       2048 kB
DirectMap4k:      124864 kB
DirectMap2M:     5117952 kB
DirectMap1G:     4194304 kB
+ python3 ../x.py test src/tools/expand-yaml-anchors
Ensuring the YAML anchors in the GitHub Actions config were expanded
Ensuring the YAML anchors in the GitHub Actions config were expanded
Building stage0 tool expand-yaml-anchors (x86_64-unknown-linux-gnu)
---
   Compiling linked-hash-map v0.5.2
   Compiling lazy_static v1.4.0
   Compiling yaml-rust v0.4.3
   Compiling quote v1.0.2
   Compiling thiserror-impl v1.0.5
   Compiling thiserror v1.0.5
   Compiling yaml-merge-keys v0.4.0
   Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
Build completed successfully in 0:00:24
+ python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu
    Finished dev [unoptimized] target(s) in 0.15s
Checking rustdoc artifacts (x86_64-unknown-linux-gnu -> i686-pc-windows-gnu)
---
    Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
    Checking rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking chalk-engine v0.11.0
    Checking rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
    Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustc_typeck/check/op.rs at line 656:
                     err.span_label(
                         ex.span,
-                        format!(
-                            "cannot apply unary operator `{}`",
-                            "cannot apply unary operator `{}`",
-                            op.as_str()
-                        ),
+                        format!("cannot apply unary operator `{}`", op.as_str()),
                     match actual.kind {
                     match actual.kind {
                         Uint(_) if op == hir::UnOp::UnNeg => {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustc_typeck/check/op.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:34
== clock drift check ==
  local time: Wed Jun 24 21:42:34 UTC 2020
  network time: Wed, 24 Jun 2020 21:42:34 GMT
  network time: Wed, 24 Jun 2020 21:42:34 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73674/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73674/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3657) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

estebank committed:
Suggest type param trait bound for binop only when appropriate

Verify that the binop trait *is* implemented for the types *if* all the
involved type parameters are replaced with fresh inferred types. When
this is the case, it means that the type parameter was indeed missing a
trait bound. If this is not the case, provide a generic `note` refering
to the type that doesn't implement the expected trait.

davidtwco reviewed (approved):
LGTM, thanks!

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 8f40dae93b3a66a9cdd0f940244da7f602618fba has been approved by `davidtwco`

<!-- @bors r=davidtwco 8f40dae93b3a66a9cdd0f940244da7f602618fba -->
<!-- homu: {"type":"Approved","sha":"8f40dae93b3a66a9cdd0f940244da7f602618fba","approver":"davidtwco"} -->
